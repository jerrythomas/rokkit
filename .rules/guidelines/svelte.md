# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
