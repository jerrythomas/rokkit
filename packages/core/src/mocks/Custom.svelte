<script>
	export let value
	export let fields
</script>

<span>{value[fields.text]}</span>
