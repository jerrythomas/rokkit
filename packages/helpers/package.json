{
  "name": "@rokkit/helpers",
  "version": "1.0.0-next.116",
  "description": "Custom matchers for vitest, mocks and simulators for testing.",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "vitest",
    "matchers",
    "validators",
    "svelte"
  ],
  "author": "Jerry Thomas<me@jerrythomas.name>",
  "license": "MIT",
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    "./mocks": "./src/mocks/index.js",
    "./matchers": "./src/matchers/index.js",
    "./simulators": "./src/simulators/index.js",
    "./components": "./src/components/index.js",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./src/index.js",
			"svelte": "./src/index.js"
    }
  },
  "scripts": {
    "prepublishOnly": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "build": "bun clean && bun prepublishOnly"
  },
  "dependencies": {
    "@vitest/expect": "^2.1.8",
    "ramda": "^0.30.1"
  }
}
