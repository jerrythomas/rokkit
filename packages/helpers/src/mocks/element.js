import { vi } from 'vitest'

/**
 * Creates an array of elements with the specified size
 *
 * @param {number} count
 * @param {number} size
 * @param {string} [prop='offsetHeight']
 * @returns {Array<Object<string, number>>}
 */
export function elementsWithSize(count, size, prop = 'offsetHeight') {
	return Array.from({ length: count }, () => ({
		[prop]: size
	}))
}

/**
 * Creates an array of elements with mixed sizes
 *
 * @param {Array<{count: number, size: number}>} data
 * @param {string} prop
 * @returns {Array<Object<string, number>>}
 */
export function mixedSizeElements(data, prop) {
	return data.reduce(
		(elements, { count, size }) => [...elements, ...elementsWithSize(count, size, prop)],
		[]
	)
}

/**
 * Creates a mock node with functions to add and remove event handlers
 *
 * @param {Array<string} events
 * @returns {{node: HTMLElement, listeners: Object<string, integer>}}
 */
export function getMockNode(events) {
	const listeners = events.reduce((acc, event) => ({ ...acc, [event]: 0 }), {})

	const node = {
		dispatchEvent: vi.fn(),
		scrollTo: vi.fn(),
		querySelectorAll: vi.fn().mockImplementation((selector) => {
			return [document.createElement(selector)]
		}),
		querySelector: vi.fn().mockImplementation((selector) => {
			return document.createElement(selector)
		}),
		addEventListener: vi.fn().mockImplementation((name) => ++listeners[name]),
		removeEventListener: vi.fn().mockImplementation((name) => --listeners[name])
	}
	return { node, listeners }
}

/**
 * @typedef {Object} NestedItem
 * @property {string} name
 * @property {string} [dataPath]
 * @property {string} [id]
 * @property {Array<NestedItem>} [children]
 */

/**
 * Creates a nested HTML element structure using the provided data
 *
 * @param {NestedItem} item
 * @returns {HTMLElement}
 */
export function createNestedElement(item) {
	const { name, dataPath, id, children } = item
	const element = document.createElement(name)

	if (dataPath) element.dataset.path = dataPath
	if (id) element.id = id

	element.scrollIntoView = vi.fn()

	if (Array.isArray(children)) {
		children.forEach((child) => element.appendChild(createNestedElement(child)))
	}

	return element
}

/**
 * Mocks the requestSubmit function for HTMLFormElement which is not implemented by JSDOM
 */
export function mockFormRequestSubmit() {
	// eslint-disable-next-line no-undef
	HTMLFormElement.prototype.requestSubmit = vi.fn(function (submitter) {
		const event = new Event('submit', {
			bubbles: true,
			cancelable: true
		})
		if (submitter) {
			event.submitter = submitter
		}
		this.dispatchEvent(event)
	})

	return true
}
