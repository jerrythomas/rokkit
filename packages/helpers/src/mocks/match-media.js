import { vi } from 'vitest'

const watchMediaQueries = []
const listeners = []

/**
 * @typedef {Object} MediaQuery
 * @property {integer} [min-width]
 * @property {integer} [max-width]
 * @property {integer} [min-height]
 * @property {integer} [max-height]
 * @property {integer} [width]
 * @property {integer} [height]
 * @property {integer} [orientation]
 * @property {integer} [aspect-ratio]
 * @property {integer} [min-aspect-ratio]
 * @property {integer} [max-aspect-ratio]
 * @property {integer} [resolution]
 * @property {integer} [min-resolution]
 * @property {integer} [max-resolution]
 * @property {integer} [scan]
 * @property {integer} [grid]
 * @property {integer} [update]
 * @property {integer} [overflow-block]
 */

/**
 * Parses a media query string into an object
 *
 * @param {string} mediaQuery
 * @returns {MediaQuery}
 */
function parseMediaQuery(mediaQuery) {
	const regex = /\(([^:]+):\s*([^)]+)\)/g
	const result = {}

	let match = null
	while ((match = regex.exec(mediaQuery)) !== null) {
		const [, property, value] = match
		result[property.trim()] = parseInt(value.trim(), 10)
	}

	return result
}

/**
 * Simulates the evaluation of a media query
 *
 * @param {MediaQuery} mediaQuery
 * @param {integer} width
 * @returns
 */
function evaluateMediaQuery(mediaQuery, width) {
	const { 'min-width': minWidth = 0, 'max-width': maxWidth = width } = mediaQuery

	return width >= minWidth && width <= maxWidth
}

/**
 * Mocks the window.matchMedia function
 * @param {string} query
 * @returns {Object}
 */
export const matchMediaMock = vi.fn().mockImplementation((query) => {
	const mediaQuery = parseMediaQuery(query)
	const handler = (width) => evaluateMediaQuery(mediaQuery, width)

	const queryObject = {
		media: query,
		matches: handler(window.innerWidth),
		addListener: vi.fn().mockImplementation((listener) => listeners.push(listener)),
		removeListener: vi
			.fn()
			.mockImplementation((listener) => listeners.splice(listeners.indexOf(listener), 1)),
		handler
	}
	watchMediaQueries.push(queryObject)
	return watchMediaQueries[watchMediaQueries.length - 1]
})

/**
 * Updates the media query matches
 * @returns {void}
 */
export function updateMedia() {
	watchMediaQueries.forEach((query) => {
		query.matches = query.handler(window.innerWidth)
	})
	listeners.forEach((listener) => listener())
}
