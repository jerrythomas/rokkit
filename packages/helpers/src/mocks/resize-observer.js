export class ResizeObserver {
	constructor(callback) {
		this.callback = callback
		this.elements = new Map()
	}

	observe(element) {
		this.elements.set(element, { contentRect: element.getBoundingClientRect() })
		// Immediately invoke the callback with the initial size
		this.callback([{ target: element, contentRect: element.getBoundingClientRect() }])
	}

	unobserve(element) {
		this.elements.delete(element)
	}

	disconnect() {
		this.elements.clear()
	}

	// Simulate a resize event
	simulateResize(element, contentRect) {
		this.callback([{ target: element, contentRect }])
	}
}
