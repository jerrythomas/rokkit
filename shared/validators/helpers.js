export function getPropertyValue(component, property) {
	return component.$$.ctx[component.$$.props[property]]
}

export function getComponentProp(component, prop) {
	return component.$$.ctx[component.$$.props[prop]]
}

export function simulateTouchSwipe(node, distance) {
	const touchStart = new Touch({
		identifier: 0,
		target: node,
		clientX: 0,
		clientY: 0
	})
	const touchEnd = new Touch({
		identifier: 0,
		target: node,
		clientX: distance.x,
		clientY: distance.y
	})
	const touchStartEvent = new TouchEvent('touchstart', {
		touches: [touchStart]
	})
	node.dispatchEvent(touchStartEvent)
	const touchEndEvent = new TouchEvent('touchend', {
		changedTouches: [touchEnd]
	})
	node.dispatchEvent(touchEndEvent)
}

export function simulateMouseSwipe(node, distance) {
	node.dispatchEvent(new MouseEvent('mouseup', { clientX: 0, clientY: 0 }))
	node.dispatchEvent(
		new MouseEvent('mousedown', { clientX: distance.x, clientY: distance.y })
	)
}
