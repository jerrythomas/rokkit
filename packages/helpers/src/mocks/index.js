import './animate'
import { ResizeObserver } from './resize-observer'

// skipcq: JS-E1004 - Needed for exposing all functions
export * from './match-media'
// skipcq: JS-E1004 - Needed for exposing all functions
export * from './element'

global.ResizeObserver = ResizeObserver
