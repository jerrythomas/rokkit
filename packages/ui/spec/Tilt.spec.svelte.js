import { describe, it, expect } from 'vitest'
import { render } from '@testing-library/svelte'
import { flushSync } from 'svelte'
import '@rokkit/helpers/mocks'
import TiltDemo from './mocks/TiltDemo.svelte'

describe('Tilt', () => {
	/*
	 * Tilt functionality changes the user experience on hover/mouse move.
	 * This test just checks the rendering only. Playwright test with snapshot would be better
	 */
	it('should render', () => {
		const props = $state({})
		const { container } = render(TiltDemo, { props })
		expect(container).toMatchSnapshot()

		props.maxRotation = 20
		flushSync()
		expect(container).toMatchSnapshot()

		props.setBrightness = true
		flushSync()
		expect(container).toMatchSnapshot()
	})
})
