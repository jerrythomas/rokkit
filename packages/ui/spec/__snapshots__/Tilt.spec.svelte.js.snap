// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tilt > should render 1`] = `
<div>
  <div
    class="[perspective:var(--perspective)] [&>*]:[transform:rotateX(var(--rotateX))_rotateY(var(--rotateY))] [&>*]:brightness-[var(--brightness)]"
    data-tilt-root=""
    style="--perspective: 600px; --rotateX: 0deg; --rotateY: 0deg; --brightness: 1;"
  >
    <span>
      Content box
    </span>
    <!---->
  </div>
  
</div>
`;

exports[`Tilt > should render 2`] = `
<div>
  <div
    class="[perspective:var(--perspective)] [&>*]:[transform:rotateX(var(--rotateX))_rotateY(var(--rotateY))] [&>*]:brightness-[var(--brightness)]"
    data-tilt-root=""
    style="--perspective: 600px; --rotateX: 0deg; --rotateY: 0deg; --brightness: 1;"
  >
    <span>
      Content box
    </span>
    <!---->
  </div>
  
</div>
`;

exports[`Tilt > should render 3`] = `
<div>
  <div
    class="[perspective:var(--perspective)] [&>*]:[transform:rotateX(var(--rotateX))_rotateY(var(--rotateY))] [&>*]:brightness-[var(--brightness)]"
    data-tilt-root=""
    style="--perspective: 600px; --rotateX: 0deg; --rotateY: 0deg; --brightness: 1;"
  >
    <span>
      Content box
    </span>
    <!---->
  </div>
  
</div>
`;
